x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/fielddata/ShapeValues.java

```diff
@@ -187,6 +200,18 @@ public double getX() throws IOException {
         public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
             throw new IllegalArgumentException("cannot write xcontent for geo_shape doc value");
         }
+
+        @Override
+        public TransportVersion getMinimalSupportedVersion() {
+            return TransportVersions.SHAPE_VALUE_SERIALIZATION_ADDED;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            byte[] bytes = this.reader.copyBytes();
+            out.writeInt(bytes.length);
```

Comment:
Suggest `out.writeVInt`, that's what we use for most array lengths.

Reply:
I would prefer to write this as a ByteReference, this arrays can  get large and I want to be able to paged if needed.

Reply:
OK. So I looked at the code for `readByteArray` vs. `readBytesReference` and `writeByteArray` vs `writeBytesReference` and they are, as Dave said, identical on the stream, and simply add an additional wrap/unwrap step for the `byte[]` inside a `BytesArray`, which just adds to GC. So I think this is not a good idea, unless we decide to use `BytesReference` as the backing structure for the `GeometryDocValueReader` for its internal triangle-tree. That seems like a completely different task out of scope for the PR and should be done in a different PR.

Reply:
That is identical over the wire is an implementation detail and can change at any moment. I insist we should use  #writeBytesReference and #readBytesReference for serialising and deserialising.

Reply:
I'm not sure that's true, any change in this area would elicit a new `TransportVersion`. We could stick with `writeByteArray()` and `readByteArray()` on both sides for now, and if we move to `writeBytesReference()` and `readBytesReference()` in a future version then we will easily be able to adapt one into the other just as with any other transport-protocol compatibility change. In practice it is extraordinarily likely that such an adapter would be a no-op, but even if it isn't it's not going to be much of an obstacle.

Reply:
I pushed a change with Ignacio's suggestion. His argument is that using BytesReference now will enable us to benefit from future changes to the serialisation planned for BytesReference, even if the current code does not yet reflect that.